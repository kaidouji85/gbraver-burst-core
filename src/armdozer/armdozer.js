// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
