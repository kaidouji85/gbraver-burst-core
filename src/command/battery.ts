/** 電池を出すコマンド */
export type BatteryCommand = {
  type: "BATTERY_COMMAND";
  battery: number;
};
