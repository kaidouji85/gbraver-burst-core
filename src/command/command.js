// @flow
import type {BatteryCommand} from "./battery";
import type {BurstCommand} from "./burst";
import type {EmptyCommand} from "./empty-command";
import type {PlayerId} from "..";
import type {PilotSkillCommand} from "./pilot-skill";

/** コマンド */
export type Command = EmptyCommand | BatteryCommand | BurstCommand | PilotSkillCommand;

/**
 * クイックコマンド
 */
export type QuickCommand = BurstCommand | PilotSkillCommand;

/**
 * 指定したコマンドをクイックコマンドにキャストする
 * キャストできない場合はnullを返す
 *
 * @param command キャストするコマンド
 * @return キャスト結果
 */
export function castQuickCommand(command: Command): ?QuickCommand {
  switch (command.type) {
    case 'BURST_COMMAND':
    case 'PILOT_SKILL_COMMAND':
      return command;
    default:
      return null;
  }
}

/**
 * プレイヤーコマンド
 *
 * @typeparam {X} コマンド
 */
export type PlayerCommandX<X> = {
  /** プレイヤーID */
  playerId: PlayerId,
  /** コマンド */
  command: X
};

/** プレイヤーコマンド */
export type PlayerCommand = PlayerCommandX<Command>;