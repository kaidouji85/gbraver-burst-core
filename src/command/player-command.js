// @flow

import type {Command} from "./command";
import type {PlayerId} from "../player/player";

export type PlayerCommand = {
  playerId: PlayerId,
  command: Command
};