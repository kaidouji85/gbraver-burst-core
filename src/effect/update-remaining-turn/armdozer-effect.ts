import type { ArmdozerEffect } from "../../state/armdozer-effect";

/**
 * アームドーザ効果の継続ターン数を更新する
 *
 * @param effect 更新前
 * @return 更新結果
 */
export function updateArmdozerEffect(effect: ArmdozerEffect): ArmdozerEffect {
  if (effect.period.type === "TurnLimit") {
    const updatedPeriod = {
      ...effect.period,
      remainingTurn: effect.period.remainingTurn - 1,
    };
    return { ...effect, period: updatedPeriod };
  }

  return effect;
}

/**
 * アームドーザ効果が継続するか否かを判定する、trueで継続する
 *
 * @param effect 判定対象
 * @return 判定結果
 */
export function isRemainArmdozerEffect(effect: ArmdozerEffect): boolean {
  if (effect.period.type === "TurnLimit") {
    return 0 < effect.period.remainingTurn;
  }

  return true;
}
