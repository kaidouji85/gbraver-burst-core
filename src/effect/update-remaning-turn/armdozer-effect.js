// @flow

import type {ArmdozerEffect} from "../..";

/**
 * アームドーザ効果の継続ターン数を更新する
 *
 * @param effect 更新前
 * @return 更新結果
 */
export function updateArmdozerEffect(effect: ArmdozerEffect): ArmdozerEffect {
  return {
    ...effect,
    remainingTurn: effect.remainingTurn - 1
  };
}

/**
 * アームドーザ効果が継続するか否かを判定する、trueで継続する
 *
 * @param effect 判定対象
 * @return 判定結果
 */
export function isRemainArmdozerEffect(effect: ArmdozerEffect): boolean {
  return 0 < effect.remainingTurn;
}