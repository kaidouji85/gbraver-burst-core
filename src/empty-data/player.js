// @flow

import type {PlayerState} from "../state/player-state";
import {EMPTY_ARMDOZER_STATE} from "./armdozer";
import {EMPTY_PILOT_STATE} from "./pilot";

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: '',
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
