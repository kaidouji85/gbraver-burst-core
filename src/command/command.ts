import { z } from "zod";

import type { BatteryCommand } from "./battery";
import { BatteryCommandSchema } from "./battery";
import type { BurstCommand } from "./burst";
import { BurstCommandSchema } from "./burst";
import type { EmptyCommand } from "./empty-command";
import { EmptyCommandSchema } from "./empty-command";
import type { PilotSkillCommand } from "./pilot-skill";
import { PilotSkillCommandSchema } from "./pilot-skill";

/** コマンド */
export type Command =
  | EmptyCommand
  | BatteryCommand
  | BurstCommand
  | PilotSkillCommand;

/** コマンド zodスキーマ */
export const CommandSchema = z.union([
  EmptyCommandSchema,
  BatteryCommandSchema,
  BurstCommandSchema,
  PilotSkillCommandSchema,
]);

/**
 * コマンドにパースする
 * @param origin パース元
 * @return パース結果、失敗した場合はnull
 */
export function parseCommand(origin: unknown): Command | null {
  const result = CommandSchema.safeParse(origin);
  return result.success ? result.data : null;
}
