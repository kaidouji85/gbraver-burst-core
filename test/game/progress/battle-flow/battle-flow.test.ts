import { BatteryCommand } from "../../../../src/command/battery";
import { EMPTY_ARMDOZER_STATE } from "../../../../src/empty/armdozer";
import { EMPTY_PLAYER_STATE } from "../../../../src/empty/player";
import { PlayerCommandX } from "../../../../src/game/command/player-command";
import { battleFlow } from "../../../../src/game/progress/battle-flow";
import { PlayerId } from "../../../../src/player/player";
import { ArmdozerEffect } from "../../../../src/state/armdozer-effect";
import { PlayerState } from "../../../../src/state/player-state";
import * as ContinuesGameData from "./battle-flow__continue-game.data";
import * as DeathData from "./battle-flow__death.data";
import * as DrawData from "./battle-flow__draw.data";

/** プレイヤー生成オプション */
type AttackerOptions = {
  /** プレイヤーID */
  playerId: PlayerId;
  /** バッテリー */
  battery: number;
  /** 適用中の効果 */
  effects: ArmdozerEffect[];
};

/**
 * 攻撃側プレイヤーを生成する
 * @param options 生成オプション
 * @returns 攻撃側プレイヤー
 */
const createAttakcer = (options: AttackerOptions): PlayerState => {
  const { playerId, battery, effects } = options;
  return {
    ...EMPTY_PLAYER_STATE,
    playerId,
    armdozer: {
      ...EMPTY_ARMDOZER_STATE,
      hp: 3000,
      maxHp: 3000,
      power: 2000,
      battery,
      maxBattery: 5,
      effects,
    },
  };
};

/**
 * バッテリーコマンドを生成する
 * @param playerId プレイヤーID
 * @param battery バッテリー値
 * @returns バッテリーコマンド
 */
const createBatteryCommand = (playerId: PlayerId, battery: number): PlayerCommandX<BatteryCommand> => ({
  playerId,
  command: { type: "BATTERY_COMMAND", battery },
});


test("戦闘したが、相手を倒しきれなかったのでゲーム続行", () => {
  const { lastState, commands } = ContinuesGameData;
  const result = battleFlow(lastState, commands);
  expect(result).toMatchSnapshot("continue-game");
});

test("攻撃で防御側のHPを0以下にした場合、ゲームが終了する", () => {
  const { lastState, commands } = DeathData;
  const result = battleFlow(lastState, commands);
  expect(result).toMatchSnapshot("death");
});

test("ダメージ反射でHPが0になった場合は引き分け", () => {
  const { lastState, commands } = DrawData;
  const result = battleFlow(lastState, commands);
  expect(result).toMatchSnapshot("draw");
});
