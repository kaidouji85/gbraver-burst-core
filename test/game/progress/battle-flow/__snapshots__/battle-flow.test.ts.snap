// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ダメージ反射でHPが0になった場合は引き分け: draw 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "attackerBattery": 2,
      "defenderBattery": 0,
      "name": "BatteryDeclaration",
      "originalBatteryOfAttacker": 2,
      "originalBatteryOfDefender": 0,
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "damage": 5000,
              "effect": "Lightning",
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "TryReflect",
            },
          ],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "isDeath": true,
      "name": "Battle",
      "result": {
        "damage": 9999,
        "name": "CriticalHit",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "damage": 5000,
              "effect": "Lightning",
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "TryReflect",
            },
          ],
          "enableBurst": true,
          "hp": -6999,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "damage": 5000,
      "damagedPlayer": "attacker",
      "effect": "Lightning",
      "isDeath": true,
      "name": "Reflect",
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": -2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "damage": 5000,
              "effect": "Lightning",
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "TryReflect",
            },
          ],
          "enableBurst": true,
          "hp": -6999,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "name": "GameEnd",
      "result": {
        "type": "EvenMatch",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": -2000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [
            {
              "damage": 5000,
              "effect": "Lightning",
              "period": {
                "remainingTurn": 1,
                "type": "TurnLimit",
              },
              "type": "TryReflect",
            },
          ],
          "enableBurst": true,
          "hp": -6999,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
]
`;

exports[`戦闘したが、相手を倒しきれなかったのでゲーム続行: continue-game 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "attackerBattery": 2,
      "defenderBattery": 1,
      "name": "BatteryDeclaration",
      "originalBatteryOfAttacker": 2,
      "originalBatteryOfDefender": 1,
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "isDeath": false,
      "name": "Battle",
      "result": {
        "damage": 2000,
        "name": "NormalHit",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 1000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "battleResult": {
        "damage": 2000,
        "name": "NormalHit",
      },
      "defender": "defender",
      "name": "RightItself",
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 1000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "endArmdozerEffects": [],
      "name": "UpdateRemainingTurn",
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 1000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "defender",
    "effect": {
      "name": "TurnChange",
      "recoverBattery": 3,
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 1000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "defender",
    "effect": {
      "name": "InputCommand",
      "players": [
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "attacker",
          "selectable": true,
        },
        {
          "command": [
            {
              "battery": 0,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 1,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 2,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 3,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 4,
              "type": "BATTERY_COMMAND",
            },
            {
              "battery": 5,
              "type": "BATTERY_COMMAND",
            },
            {
              "type": "BURST_COMMAND",
            },
            {
              "type": "PILOT_SKILL_COMMAND",
            },
          ],
          "playerId": "defender",
          "selectable": true,
        },
      ],
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 5,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 1000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
]
`;

exports[`攻撃で防御側のHPを0以下にした場合、ゲームが終了する: death 1`] = `
[
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "attackerBattery": 2,
      "defenderBattery": 1,
      "name": "BatteryDeclaration",
      "originalBatteryOfAttacker": 2,
      "originalBatteryOfDefender": 1,
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 100,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "attacker": "attacker",
      "isDeath": true,
      "name": "Battle",
      "result": {
        "damage": 2000,
        "name": "NormalHit",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": -1900,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
  {
    "activePlayerId": "attacker",
    "effect": {
      "name": "GameEnd",
      "result": {
        "type": "GameOver",
        "winner": "attacker",
      },
    },
    "players": [
      {
        "armdozer": {
          "battery": 2,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": 3000,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "attacker",
      },
      {
        "armdozer": {
          "battery": 4,
          "burst": {
            "recoverBattery": 0,
            "turnStartBatteryCorrect": 0,
            "type": "RecoverBattery",
          },
          "effects": [],
          "enableBurst": true,
          "hp": -1900,
          "id": "id",
          "maxBattery": 5,
          "maxHp": 3000,
          "name": "name",
          "power": 2000,
          "speed": 2000,
        },
        "pilot": {
          "enableSkill": true,
          "id": "EMPTY_PLAYER",
          "name": "名無し",
          "skill": {
            "recoverBattery": 2,
            "type": "RecoverBatterySkill",
          },
        },
        "playerId": "defender",
      },
    ],
  },
]
`;
